jep/0000/README.adoc

```diff
@@ -0,0 +1,267 @@
+
+== Motivation
+
+Jenkins uses `pageHeader.jelly` file to specify the content of the header. Although it is valid, if a user wants to modify the header to perform some branding operation, Jenkins header branding capabilities are limited.
+
+Some plugins, like: https://plugins.jenkins.io/simple-theme-plugin/[simple-theme-plugin], allow Jenkins users to customize some parts like CSS and/or Javascript.
+On the other hand, if a Jenkins user wants to customize/modify some additional business functionality on some menus and search bar, then there is no approach beyond updating/overriding `pageHeader.jelly` from the Jenkins core, which would be a problem on updating the instance due to conflicts.
+In addition, if the user wants to customize these behaviors it would be good to have those features as REST endpoints.
+
+Also, other alternatives have been evaluated as workaround like the https://github.com/stephenc/diffpatch-maven-plugin[diffpatch-maven-plugin] with no satisfactory results. See the _Reasoning_ section for further details.
+
+So, this approach is not only about providing UI capabilities but also about providing extra functionality and better integration.
+
+== Reasoning
+
+Let's consider the following dummy example to illustrate reasoning on why particular design decisions were made and also why current approaches were discarded.
+
+> A Jenkins user wants to modify its current Jenkins instance header to provide a configurable message (default: `Hello World!`) with the account username of the logged user, as well as two (why not?) search bars
+
+Some options exist to perform parts of the required actions mentioned above:
+
+* Update the `pageHeader.jelly` content of their Jenkins instance.
+  Discarded to not require overriding/updating original Jenkins core source code.
+* Use https://plugins.jenkins.io/simple-theme-plugin/[simple-theme-plugin]. It could help us to modify some CSS and Javascript elements, and could be used to reach our goals, but it would be so hacky and will not be able to retrieve programatically the configurable message to be included with the username of the logged user.
+* Use https://github.com/stephenc/diffpatch-maven-plugin[diffpatch-maven-plugin] to override the content of the header using patches. It would help us have a second search box, but not to have the configurable message because it will only help for static content.
+
+Given existing approaches do not fulfill this example, we will explore the alternative of a Jenkins user that wants to update the Jenkins header through an ad-hoc Jenkins plugin that follows the principles provided in the Specification section.
+
+The following source code is provided just for illustrative purposes on reasoning why particular design decisions were made. It may differ from the final implementation (please go to Reference Implementation section to see the final proposed changes in source code)
```

Comment:
Not sure why this section exists. Just delete from here to **Backwards Compatibility** IMO.

Reply:
As mentioned in the document:

> The following source code is provided just for illustrative purposes on the reasoning why particular design decisions were made. It may differ from the final implementation.